src/OrbitGl/LineGraphTrack.cpp

```diff
@@ -0,0 +1,117 @@
+                         this->layout_->GetTrackBottomMargin() - this->GetLegendHeight();
+  float base_y = this->pos_[1] - this->size_[1] + this->layout_->GetTrackBottomMargin();
+  double min = this->GetGraphMinValue();
+  double inverse_value_range = this->GetGraphInverseValueRange();
+  std::array<float, Dimension> normalized_values;
+  std::transform(values.begin(), values.end(), normalized_values.begin(),
+                 [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  if (Dimension == 1) return base_y + normalized_values[0] * content_height;
+  return base_y + content_height / 2.0f;
+}
+
+template <size_t Dimension>
+void LineGraphTrack<Dimension>::DrawSeries(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                           float z) {
+  auto entries_affected_range_result =
+      this->series_.GetEntriesAffectedByTimeRange(min_tick, max_tick);
+  if (!entries_affected_range_result.has_value()) return;
+
+  typename MultivariateTimeSeries<Dimension>::Range& entries{entries_affected_range_result.value()};
+
+  double min = this->GetGraphMinValue();
+  double inverse_value_range = this->GetGraphInverseValueRange();
+
+  auto current_iterator = entries.begin;
+  uint64_t current_time = current_iterator->first;
+  std::array<float, Dimension> current_normalized_values;
+  std::transform(current_iterator->second.begin(), current_iterator->second.end(),
+                 current_normalized_values.begin(), [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  while (current_iterator != entries.end) {
+    auto next_iterator = std::next(current_iterator);
+    uint64_t next_time = next_iterator->first;
+    std::array<float, Dimension> next_normalized_values;
+    std::transform(next_iterator->second.begin(), next_iterator->second.end(),
+                   next_normalized_values.begin(), [min, inverse_value_range](double value) {
+                     return static_cast<float>((value - min) * inverse_value_range);
+                   });
+    bool is_last = next_time >= max_tick;
+
+    DrawSingleSeriesEntry(batcher, current_time, next_time, current_normalized_values,
+                          next_normalized_values, z, is_last);
+
+    current_iterator = next_iterator;
+    current_time = next_time;
+    current_normalized_values = next_normalized_values;
+  }
+
+  if (current_time < max_tick) {
+    DrawSingleSeriesEntry(batcher, current_time, max_tick, current_normalized_values,
+                          current_normalized_values, z, true);
+  }
+}
+
+static void DrawSquareDot(Batcher* batcher, Vec2 center, float radius, float z,
+                          const Color& color) {
+  Vec2 position(center[0] - radius, center[1] - radius);
+  Vec2 size(2 * radius, 2 * radius);
```

Comment:
As discussed, I would like to see if drawing the square with an odd number of pixels could make it nicer

Reply:
We are considering increasing the line width to 2 so to fix the missing horizontal line issue~ and the square size will be increased accordingly to guarantee that lines are drawn at the centre of the square dots. 

I will update this part in another PR 😃 Thanks for pointing this out!